torch/distributed/_spmd/data_parallel.py

```diff
@@ -0,0 +1,731 @@
+    """
+
+    DEFAULT = 0
+    REPLICATE = 1
+    FULLY_SHARD = 2
+
+
+class NodeType(Enum):
+    """
+    NodeType is a enum that records the type of the tensors in the graph.
+    This is used to determine the data parallel strategy.
+    """
+
+    PARAM = 0
+    ACT = 1
+    GRAD = 2
+    STATE = 3
+    NON_TENSOR = 4  # NON_TENSOR is to tag non tensor node (i.e. graph output)
+
+
+@dataclass
+class DataParallelStrategy(StrategyList):
+    """
+    DataParallelStrategy is a special case of StrategyList that only records
+    the "data parallel style" placement strategy for each fx Node.
+    """
+
+    node_type: NodeType
+    reduction_over_batch: bool
+
+    def __init__(
+        self,
+        node_type: NodeType,
+        startegy_list: StrategyList,
+        reduction_over_batch: bool = False,
+    ):
+        super().__init__(startegy_list)
+        self.node_type = node_type
+        self.reduction_over_batch = reduction_over_batch
+
+    def __str__(self) -> str:
+        return f"type: {self.node_type}, {super().__str__()}"
+
+
+class BatchDimAnalyzer(object):
+    """
+    This class is used to analyze the batch dimension of each tensor/node in the
+    graph. We need to know the batch dimension of each tensor/node so that we know
+    exactly the sharding layout of intermediate tensors.
+
+    We possibly should use symbolic shapes to track the batch dimension, but this
+    needs to happen with dynamo, as dynamo is the only place we can mark a single
+    dimension. For now, we just use the batch dimension of the first input tensor
+    as the hint to track the batch dimension of all tensors/nodes in the graph.
+    """
+
+    def __init__(self, batch_dim: int = 0) -> None:
+        self.batch_dim = batch_dim
+
+        if batch_dim != 0:
+            # TODO: see if this make sense or not
+            raise RuntimeError("Data Parallel only supports batch dim on dimension 0!")
+
+        self.batch_dim_map: Dict[fx.Node, int] = {}
+        # batch dim size is used to track the batch dim size of the input tensor
+        self.batch_dim_size = -1
+
+    def init_batch_dim_size(self, batch_dim_size: int) -> None:
+        """
+        initialize batch dim size base on the first input batch size
+        """
+        if self.batch_dim_size != -1 and self.batch_dim_size != batch_dim_size:
+            raise RuntimeError(
+                f"batch dim size is already initialized! "
+                f"Found new batch size: {batch_dim_size} not "
+                f"matching existing batch dim size: {self.batch_dim_size}!"
+            )
+        self.batch_dim_size = batch_dim_size
+
+    def set_batch_dim(self, node: fx.Node, batch_dim: int) -> None:
+        self.batch_dim_map[node] = batch_dim
+
+    def get_batch_dim(self, node: fx.Node) -> int:
+        if node not in self.batch_dim_map:
+            raise RuntimeError(f"batch dim analysis failed on node: {node}!")
```

Comment:
Do we ever expect this to fail? Or, is this indicative of an internal implementation issue with `BatchDimAnalyzer`?

Reply:
This is most likely just an indicative of an internal implementation issue so that we can fix it.